### PR TITLE
cks: k8s cluster on vpc tier

### LINKED ISSUE #6723: Kubernetes Clusters cannot be deployed in VPCs
##### ISSUE TYPE

 * Bug Report
 * Improvement Request
 * Documentation Report

##### COMPONENT NAME

~~~
Kubernetes Provider and VPC
~~~

##### CLOUDSTACK VERSION

~~~
4.17.0.1
~~~

##### CONFIGURATION

Advanced Networking

##### OS / ENVIRONMENT

Ubuntu based hosts using KVM Hypervisor

##### SUMMARY

According to the documentation [Cloudstack Kubernetes Service](https://docs.cloudstack.apache.org/en/latest/plugins/cloudstack-kubernetes-service.html#kubernetes-clusters) a Kubernetes Cluster can be deployed in VPC Networks.

Unfortunately, creating a Kubernetes Cluster in a VPC (VPC enabled networks) is failing, because of the missing "Firewall Service" in all networks being VPC enabled.

Creating a NetworkOffering with VPC support does not allow to enable the "Firewall Service" via the provider VpcVirtualRouter, which makes sense, because VPC enabled network offerings use the "Network ACL Service".

Therefore i think either the documentation is not specific enough to exclude VPC, or the Kubernetes Service might require additional functionality to support VPC enabled networks and their "Network ACL Service" instead of the "Firewall Service".
Or i miss something obvious / misunderstand the documentation.

##### STEPS TO REPRODUCE

~~~
- Create a VPC if not existing
- Create a Tier within the VPC if not existing, for example using the DefaultIsolatedNetworkOfferingForVpcNetworks Network Offering
- Try to create a Kubernetes Cluster selecting either the previously created VPC Tier network or any other VPC Tier network
~~~

Cloudstack will complain about the missing "Firewall Service" of the chosen network.

##### EXPECTED RESULTS

~~~
Either the documentation should not mention the possibility to deploy Kubernetes Clusters in VPCs, or the Kubernetes Cluster Deployment should be able to handle "Network ACL Service" in addition the the currently supported "Firewall Service"
~~~

##### ACTUAL RESULTS

~~~
Cloudstack will complain about the missing "Firewall Service" of the chosen network when trying to deploy a Kubernetes Cluster in any kind of VPC enabled network.
~~~

It would be great to see Kubernetes Clusters working in VPCs, because of the richer feature set the virtual router provides on VPCs, for example like easy possibility for Tiering / subnet separation and the possibility of Private Gateways with static routes.

### PR DESCRIPTION
**Pull request by shwstppr**
### Description

Fixes #6723 

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
Great job ! @shwstppr

File: plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/KubernetesClusterManagerImpl.java [hunk 1]
```diff
@@ -352,56 +352,57 @@ public VMTemplateVO getKubernetesServiceTemplate(DataCenter dataCenter, Hypervis
...
-                    LOGGER.debug("Network rule : " + startPort + " " + endPort);
-                }
-                if (startPort <= KubernetesClusterActionWorker.CLUSTER_API_PORT && KubernetesClusterActionWorker.CLUSTER_API_PORT <= endPort) {
-                    throw new InvalidParameterValueException(String.format("Network ID: %s has conflicting port forwarding rules to provision Kubernetes cluster for API access", network.getUuid()));
-                }
-                if (startPort <= KubernetesClusterActionWorker.CLUSTER_NODES_DEFAULT_START_SSH_PORT && KubernetesClusterActionWorker.CLUSTER_NODES_DEFAULT_START_SSH_PORT + clusterTotalNodeCount <= endPort) {
-                    throw new InvalidParameterValueException(String.format("Network ID: %s has conflicting port forwarding rules to provision Kubernetes cluster for node VM SSH access", network.getUuid()));
-                }
+        }
+        rules = firewallRulesDao.listByIpAndPurposeAndNotRevoked(sourceNatIp.getId(), FirewallRule.Purpose.PortForwarding);
```
Comment (Pearl1594) [inline plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/KubernetesClusterManagerImpl.java:380]: Can we extract L380 - L 391 , as it seems similar to L366 - L377?

File: plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/actionworkers/KubernetesClusterDestroyWorker.java [hunk 2]
```diff
@@ -258,6 +296,14 @@ public boolean destroy() throws CloudRuntimeException {
                     updateKubernetesClusterEntryForGC();
                     throw new CloudRuntimeException(msg, e);
                 }
+                try {
+                    releaseVpcTierPublicIpIfNeeded();
+                } catch (InsufficientAddressCapacityException e) {
```
Comment (Pearl1594) [inline plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/actionworkers/KubernetesClusterDestroyWorker.java:301]: Not sure if Insufficient address capacity would be the most appropriate exception type?

File: plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/actionworkers/KubernetesClusterResourceModifierActionWorker.java [hunk 3]
```diff
@@ -447,52 +447,51 @@ protected void provisionFirewallRules(final IpAddress publicIp, final Account ac
...
+                            DEFAULT_SSH_PORT, DEFAULT_SSH_PORT,
+                            "tcp", networkId, accountId, domainId, vmId);
+            newRule.setDisplay(true);
+            newRule.setState(FirewallRule.State.Add);
+            newRule = portForwardingRulesDao.persist(newRule);
+            return newRule;
+        });
+        rulesService.applyPortForwardingRules(publicIp.getId(), account);
+        if (LOGGER.isInfoEnabled()) {
+            LOGGER.info(String.format("Provisioned SSH port forwarding rule: %s from port %d to 22 on %s to the VM IP : %s in Kubernetes cluster : %s", pfRule.getUuid(), sourcePort, publicIp.getAddress().addr(), vmIp.toString(), kubernetesCluster.getName()));
```
Comment (Pearl1594) [inline plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/actionworkers/KubernetesClusterResourceModifierActionWorker.java:472]: nit: we could probably replace the hard coded end port (22) to the constant DEFAULT_SSH_PORT to be safer..

**Review by Pearl1594 — COMMENTED** [review]
Nice stuff @shwstppr. I realize the PR is in draft, just added a few small comments on first pass of review. Thanks!

**Comment by shwstppr** [comment]
@Pearl1594 cluster deployment isn't working as of now that's why in draft. Not sure why, need to debug. Will address your comments alongwith fixes.

**Comment by weizhouapache** [comment]
> @Pearl1594 cluster deployment isn't working as of now that's why in draft. Not sure why, need to debug. Will address your comments alongwith fixes.

@shwstppr 
would it require some changes in ccm ?
https://github.com/apache/cloudstack-kubernetes-provider

**Comment by shwstppr** [comment]
@weizhouapache don't think so as it doesn't play a role during the cluster deployment. Last I tested it seems to be no setting up services defined in the YAML so could be something related to cloud-init or some issue with MS and node VM connectivity. I've redone my setup and will debug. Will keep you guys updated

File: plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/actionworkers/KubernetesClusterDestroyWorker.java (hunk 2 above)
Comment (shwstppr) [reply plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/actionworkers/KubernetesClusterDestroyWorker.java:301]: @Pearl1594 that is what NetworkService.releaseIpAddress throws

File: plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/actionworkers/KubernetesClusterResourceModifierActionWorker.java (hunk 3 above)
Comment (shwstppr) [reply plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/actionworkers/KubernetesClusterResourceModifierActionWorker.java]: Done

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@weizhouapache @Pearl1594 update: cluster deployment is working fine now in my tests. Will do some more tests and will add some unit tests before marking this ready for review

**Comment by weizhouapache** [comment]
great @shwstppr 

can you add a smoke test for it ?

**Comment by shwstppr** [comment]
> great @shwstppr
> 
> can you add a smoke test for it ?

@weizhouapache should I add a smoke test in the existing `test/integration/smoke/test_kubernetes_clusters.py`? Asking this because those tests already take a considerable amount of time but I guess we don't have many options?

**Comment by weizhouapache** [comment]
> > great @shwstppr
> > can you add a smoke test for it ?
> 
> @weizhouapache should I add a smoke test in the existing `test/integration/smoke/test_kubernetes_clusters.py`? Asking this because those tests already take a considerable amount of time but I guess we don't have many options?

@shwstppr 
14 hours and 15 hours make small difference for me :-D

we do not need to test the full life cycle of cks cluster on vpc tier.
I am ok with the smoke test which create and destroy cks cluster. it might take 10~20 mins ?

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan test

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan test

**Comment by rohityadavcloud** [comment]
@blueorangutan test

**Comment by rohityadavcloud** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan test

**Comment by shwstppr** [comment]
@blueorangutan LLtest

**Comment by shwstppr** [comment]
@blueorangutan LLtest

**Review by vladimirpetrov — APPROVED** [review]
LGTM based on manual testing.

File: plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/actionworkers/KubernetesClusterDestroyWorker.java (hunk 2 above)
Comment (Pearl1594) [reply plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/actionworkers/KubernetesClusterDestroyWorker.java:298]: Got It @shwstppr. Thanks

**Review by Pearl1594 — APPROVED** [review]
CLGTM

**Comment by shwstppr** [comment]
@blueorangutan LLpackage

**Comment by shwstppr** [comment]
@blueorangutan LLpackage

**Comment by shwstppr** [comment]
@blueorangutan LLtest

**Comment by DaanHoogland** [comment]
I think we can merge @shwstppr , but let me do another build/test-round.
@blueorangutan package

**Pull request merged**